sdk/storage/azure-storage-files-shares/src/share_directory_client.cpp

```diff
@@ -167,16 +167,77 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
     {
       protocolLayerOptions.FilePermission = std::string(c_FileInheritPermission);
     }
-    return Details::ShareRestClient::Directory::Create(
+    auto result = Details::ShareRestClient::Directory::Create(
         m_shareDirectoryUri, *m_pipeline, options.Context, protocolLayerOptions);
+    Models::CreateDirectoryResult ret;
+    ret.Created = true;
+    ret.ETag = std::move(result->ETag);
+    ret.SmbProperties.Attributes = Details::FileAttributesListFromString(result->FileAttributes);
+    ret.SmbProperties.CreatedOn = std::move(result->FileCreatedOn);
+    ret.SmbProperties.LastWrittenOn = std::move(result->FileLastWrittenOn);
+    ret.SmbProperties.PermissionKey = std::move(result->FilePermissionKey);
+    ret.FileChangedOn = std::move(result->FileChangedOn);
+    ret.FileId = std::move(result->FileId);
+    ret.FileParentId = std::move(result->FileParentId);
+    ret.IsServerEncrypted = result->IsServerEncrypted;
+    ret.LastModified = std::move(result->LastModified);
+
+    return Azure::Core::Response<Models::CreateDirectoryResult>(
+        std::move(ret), result.ExtractRawResponse());
+  }
+
+  Azure::Core::Response<Models::CreateDirectoryResult> ShareDirectoryClient::CreateIfNotExists(
+      const CreateDirectoryOptions& options) const
+
+  {
+    try
+    {
+      return Create(options);
+    }
+    catch (StorageException& e)
+    {
+      if (e.ErrorCode == Details::ResourceAlreadyExists)
+      {
+        Models::CreateDirectoryResult ret;
+        ret.Created = false;
+        return Azure::Core::Response<Models::CreateDirectoryResult>(
+            std::move(ret), std::move(e.RawResponse));
+      }
+      throw;
+    }
   }
 
   Azure::Core::Response<Models::DeleteDirectoryResult> ShareDirectoryClient::Delete(
       const DeleteDirectoryOptions& options) const
   {
     auto protocolLayerOptions = Details::ShareRestClient::Directory::DeleteOptions();
-    return Details::ShareRestClient::Directory::Delete(
+    auto result = Details::ShareRestClient::Directory::Delete(
         m_shareDirectoryUri, *m_pipeline, options.Context, protocolLayerOptions);
+    Models::DeleteDirectoryResult ret;
+    ret.Deleted = true;
+    return Azure::Core::Response<Models::DeleteDirectoryResult>(
+        std::move(ret), result.ExtractRawResponse());
+  }
+
+  Azure::Core::Response<Models::DeleteDirectoryResult> ShareDirectoryClient::DeleteIfExists(
+      const DeleteDirectoryOptions& options) const
+  {
+    try
+    {
+      return Delete(options);
+    }
+    catch (StorageException& e)
+    {
+      if (e.ErrorCode == Details::ShareNotFound || e.ErrorCode == Storage::Details::ParentNotFound
+          || e.ErrorCode == Storage::Details::ResourceNotFound)
```

Comment:
When will server return `ParentNotFound` or `ResourceNotFound`?

Reply:
How are they different from `ShareNotFound`?

Reply:
`ParentNotFound` when parent directory is not there, and `ResourceNotFound` when the directory to be deleted is not there. Either the share does not exist, or the parent directory does not exist or the resource does not exist should not be reporting error here.